### PR TITLE
Handles obsolete PDB files for the XML and MMCIF formats.

### LINKED ISSUE #3988: Extend PDBList obsolete handling to mmCIF and mmtf formats as well as assemblies
The current code in `PDBList.py` to move obsolete files expects a `.ent` extension, which is reserved for PDB-formatted files. We should extend this to depend on the file_format of choice (e,g. mmCIF, mmtf) and also to assemblies (if the original entry is labelled obsolete). Relevant code is [here](https://github.com/JoaoRodrigues/biopython/blob/4d7042ed1185505691c7275c50bb539ef871211a/Bio/PDB/PDBList.py#L389).

Fixing this involves keeping track of which extensions are used for which file formats, probably by encoding a dictionary to map file_format to extension, and then retrieving the extension based on the file_format provided to `update_pdb`.

### PR DESCRIPTION
**Pull request by JulianMaurin**
Hello :wave: 

This RP is intended to resolve issue #3988.

One more time I deviated from the proposed implementation favoring an enum rather than a dict.

I took advantage of this PR to perform the first wave of refactoring (and typing) on most of the `PDBList` methods.

Reviewing commit by commit is highly recommended. There are many but they are very small and (as much as possible) are single purpose.
Thanks :bow: 

ps: I will conflict myself with #4012, but no problem it will be quite simple.
______

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

______

Closes #3988

### CODE REVIEW

**Comment by JoaoRodrigues** [comment]
Hi Julian,

Thank you for your contributions! I will only be able to look at them on
Tuesday, but they look very good at a glance! Enum is preferred indeed,
good choice.

One important thing, keep in mind the work on #3986 when refactoring!

Cheers,

João

A domingo, 31/07/2022, 09:22, Julian Maurin ***@***.***>
escreveu:

File: Bio/PDB/PDBList.py [hunk 1]
```diff
@@ -435,52 +443,62 @@ def update_pdb(self, file_format=None):
...
             try:
-                self.retrieve_pdb_file(pdb_code, file_format=file_format)
-            except Exception:
-                print(f"error {pdb_code}\n")
-                # you can insert here some more log notes that
-                # something has gone wrong.
+                self.retrieve_pdb_file(
+                    pdb_code,
+                    file_format=file_format_enum,
+                    overwrite=pdb_code in modified,
```
Comment (JulianMaurin) [inline Bio/PDB/PDBList.py:465]: Notice that the PDB files tagged as `modified` were not overridden.
Regarding the context, I think it's what we want, wdyt ?

**Comment by JulianMaurin** [comment]
Hello @JoaoRodrigues !
Yes, It was a lot of fun to work on this, I still have a lot of ideas to improve this module.

And also thank you for bringing this PR to my attention.
Indeed it is better to wait until it is merged since it is already well advanced. I offered to help with the implementation of the tests (which in my opinion are a must-have).

Maybe I should split this PR in two. Within the first one the refactoring part that doesn't touch the code in question, wdyt?
In any case, let's focus on #3986 and #4012 in the first instance, I'll keep this PR in draft 👌 

Cheers,
Julian

**Comment by JulianMaurin** [comment]
Closing this pull request to avoid noise in the PRs list
Waiting for move on others PRs to open it back ([see](https://github.com/biopython/biopython/pull/4014#issuecomment-1200440188))

**Pull request closed without merging**